pandas/core/indexes/base.py

```diff
@@ -5516,65 +5549,24 @@ def _add_logical_methods(cls):
         >>> pd.Index([0, 0, 0]).any()
         False
         """
-        )
-
-        _index_shared_docs["index_any"] = dedent(
-            """
-
-        See Also
-        --------
-        Index.all : Return whether all elements are True.
-        Series.all : Return whether all elements are True.
+        # FIXME: docstr inaccurate, args/kwargs not passed
 
-        Notes
-        -----
-        Not a Number (NaN), positive infinity and negative infinity
-        evaluate to True because these are not equal to zero.
+        self._maybe_disable_logical_methods("all")
+        return np.all(self.values)
 
-        Examples
-        --------
-        >>> index = pd.Index([0, 1, 2])
-        >>> index.any()
-        True
-
-        >>> index = pd.Index([0, 0, 0])
-        >>> index.any()
-        False
+    def _maybe_disable_logical_methods(self, opname: str_t):
         """
-        )
-
-        def _make_logical_function(name: str_t, desc: str_t, f):
-            @Substitution(outname=name, desc=desc)
-            @Appender(_index_shared_docs["index_" + name])
-            @Appender(_doc)
-            def logical_func(self, *args, **kwargs):
-                result = f(self.values)
-                if (
-                    isinstance(result, (np.ndarray, ABCSeries, Index))
-                    and result.ndim == 0
-                ):
-                    # return NumPy type
-                    return result.dtype.type(result.item())
-                else:  # pragma: no cover
-                    return result
-
-            logical_func.__name__ = name
-            return logical_func
-
-        cls.all = _make_logical_function(
-            "all", "Return whether all elements are True.", np.all
-        )
-        cls.any = _make_logical_function(
-            "any", "Return whether any element is True.", np.any
-        )
-
-    @classmethod
-    def _add_logical_methods_disabled(cls):
+        raise if this Index subclass does not support any or all.
         """
-        Add in logical methods to disable.
-        """
-        cls.all = make_invalid_op("all")
-        cls.any = make_invalid_op("any")
+        if (
+            isinstance(self, ABCMultiIndex)
+            or needs_i8_conversion(self.dtype)
+            or is_interval_dtype(self.dtype)
+            or is_categorical_dtype(self.dtype)
+            or is_float_dtype(self.dtype)
+        ):
+            # This call will raise
+            make_invalid_op(opname)(self)
```

Comment:
could this be done on the subclasses. more repetitive but would be easier to follow I think.

although, does disabling on subclasses obey LSP?

maybe a mixin that classes can opt into, to ensure that the base class or other code does not rely on any/all being available.

This would mean a base class for Index, to allow any/all on object Indexes?

Reply:
ATM we disable on subclasses (though by calling `cls.add_disabled_whatever()`), and i thought it was clearer to do this in just one place, but if people disagree im fine with going the other direction.

I don't think it's _that_ big a deal because I expect to change it before long since, as mentioned in the OP, i think several of the currently-disabled methods should be enabled.

Reply:
I'm OK with perhaps revisiting after some of the currently-disabled methods are enabled.